### PR TITLE
MAINT Use dependency injection for JS PackageManager

### PR DESCRIPTION
**Pull request by ryanking13**
### Description

Updates PackageManager class to accept `API` and `Module` as a parameter instead of using a globally defined variable. This enables unit test for Package Manager class as we can inject mock object when testing.

### Checklists

- [x] Add / update tests

### CODE REVIEW

**Comment by ryanking13** [comment]
@hoodmane 

I am not sure how should I interpret the snapshot error:

```
Error: Failed to get value at index 7
    at API.makeSnapshot (/root/repo/dist/pyodide.asm.js:10:103507)
    at Object.makeMemorySnapshot (/root/repo/dist/pyodide.asm.js:10:110080)
    at file:///root/repo/dist/makesnap.mjs:8:47
```

**Comment by hoodmane** [comment]
I'll take a look soon.

**Comment by ryanking13** [comment]
Thanks!

File: src/js/dynload.ts [hunk 1]
```diff
@@ -201,5 +201,7 @@ export async function loadDynlibsFromPackage(
   }
 }
 
-API.loadDynlib = loadDynlib;
-API.loadDynlibsFromPackage = loadDynlibsFromPackage;
+if (typeof globalThis.API !== "undefined") {
```
Comment (hoodmane) [inline src/js/dynload.ts:204]: You'd usually either want the condition to be `typeof API !== "undefined"` or `globalThis.API != undefined`. Or maybe just `if (globalThis.API)`. 

But why would `globalThis.API` be defined? Is this for testing? If so could you add a comment?

File: src/js/load-package.ts [hunk 2]
```diff
@@ -165,7 +192,14 @@ class PackageManager {
     const loadedPackageData = new Set<InternalPackageData>();
     const messageCallback = options.messageCallback || console.log;
     const errorCallback = options.errorCallback || console.error;
-    if (names instanceof PyProxy) {
+
+    // originally, this condition was "names instanceof PyProxy",
+    // but it is changed to check names.toJs so that we can use type-only import for PyProxy and remove side effects.
+    // this change is required to run unit tests against this file, when global API or Module is not available.
+    // TODO: remove side effects from pyproxy.ts so that we can directly import PyProxy
```
Comment (hoodmane) [inline src/js/load-package.ts:199]: Huh, I wonder how hard this will be.

Comment (hoodmane) [reply src/js/load-package.ts:199]: But this replacement code isn't awful to be honest.

File: src/js/load-package.ts [hunk 3]
```diff
@@ -165,7 +192,14 @@ class PackageManager {
...
+    // @ts-ignore
+    if (typeof names.toJs === "function") {
+      // @ts-ignore
       names = names.toJs();
     }
     if (!Array.isArray(names)) {
```
Comment (hoodmane) [inline src/js/load-package.ts:205]: I wonder if we should check here if `typeof names  === "string"` and then we wouldn't need the type assertion.

**Review by hoodmane — APPROVED** [review]
Looks reasonable to me.

File: src/js/load-package.ts (hunk 2 above)
Comment (ryanking13) [reply src/js/load-package.ts:199]: I think we can fix it by cleaning up the pyproxy.ts code a bit, but it didn't work out easily the first few times I tried.

File: src/js/load-package.ts (hunk 3 above)
Comment (ryanking13) [reply src/js/load-package.ts:205]: Thanks for the suggestion. I tried that change but TS was complaining about the PyProxy type as the type is not narrowed enough.

File: src/js/dynload.ts (hunk 1 above)
Comment (ryanking13) [reply src/js/dynload.ts]: `typeof API !== "undefined"` worked. Thanks!

File: src/js/snapshot.ts [hunk 4]
```diff
@@ -12,6 +12,7 @@ export function getExpectedKeys() {
     API,
     scheduleCallback,
     API,
+    {},
```
Comment (hoodmane) [inline src/js/snapshot.ts:15]: Confusing, this just undoes my change? You have any idea which of your subsequent changes made it go back? We probably need to see if we can come up with a better way to keep this list up to date...

Comment (ryanking13) [reply src/js/snapshot.ts:15]: > Confusing, this just undoes my change? You have any idea which of your subsequent changes made it go back? We probably need to see if we can come up with a better way to keep this list up to date...

I think changing `typeof globalThis.API` ==> `typeof API` made it go back for some reason... I have no idea way.

**Pull request merged**